### PR TITLE
T199: using rfft and window function correction (considering edge cases)

### PR DESCRIPTION
**Pull request by cbender98**
…tually always used values in Hz (in contrast to the wrong comments of the functions)

debugged unsymmetric window, edge case and rounding of cutoffs in tukey bandpass

added tests for testing edge cases (no filtering, filtering only one frequency)

### CODE REVIEW

**Review by TomTomRixRix — CHANGES_REQUESTED** [review]
I haven't run the code and tested it yet, but just from looking at the diff I've inserted some comments. You might probably also update the PR title and add the current text into the description as the PR is not just about the "in_Hz" addition but mostly about the rfft and window function correction.

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py [hunk 1]
```diff
@@ -185,24 +185,26 @@ def tukey_bandpass_filtering(data: np.ndarray, time_spacing_in_ms: float = None,
...
-    win = tukey(large_index - small_index, alpha=tukey_alpha)
-    window = np.zeros(target_size)
-    window[small_index:large_index] = win
-    if small_index == 0:
-        small_index = 1
-        win = win[:-1]
-    window[-large_index:-small_index] = win
+    win = tukey(large_index - small_index + 1, alpha=tukey_alpha)
+    window = np.zeros_like(frequencies)
+    window[small_index:large_index+1] = win
```
[the remark this answers is not in the document]
Comment (cbender98) [reply simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:202]: since indexing [:INDEX] does not include the value at INDEX, it should be correct.

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py [hunk 2]
```diff
@@ -185,24 +185,26 @@ def tukey_bandpass_filtering(data: np.ndarray, time_spacing_in_ms: float = None,
...
 
     # construct bandpass filter given the cutoff values with tukey window in negative and positive frequencies
-    win = tukey(large_index - small_index, alpha=tukey_alpha)
-    window = np.zeros(target_size)
-    window[small_index:large_index] = win
-    if small_index == 0:
-        small_index = 1
-        win = win[:-1]
-    window[-large_index:-small_index] = win
+    win = tukey(large_index - small_index + 1, alpha=tukey_alpha)
```
[the remark this answers is not in the document]
Comment (cbender98) [reply simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:200]: when i tested the edge cases (no filtering at all, exactly one frequency filtering), i came up with this solution:
no filtering at all:
then small index would be zero and large index would be the last index of the array, however the len of the time series data is not just large_index - small_index but large_index - small_index +1. if you want to apply a window with 1s for the whole array it has to be of the same size

flooring the large_index should correspond to neglecting all frequencies above the the given lowpass cutoff (which makes sense for me and what actually was in the comment of the old version but was not correctly implemented, imo)

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py [hunk 3]
```diff
@@ -185,24 +185,26 @@ def tukey_bandpass_filtering(data: np.ndarray, time_spacing_in_ms: float = None,
 
         data = zoom(data, zoom_factors, order=order, mode=mode)
 
+        time_spacing_in_ms *= resampling_factor
+
+    # array of frequencies in Hz corrsponding to rfft and irfft method
```
[the remark this answers is not in the document]
Comment (cbender98) [reply simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:190]: added this in the description of the function (line 154)

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py [hunk 4]
```diff
@@ -185,24 +185,26 @@ def tukey_bandpass_filtering(data: np.ndarray, time_spacing_in_ms: float = None,
...
     # compute closest indices for cutoff frequencies, limited by the Nyquist frequency
-    single_voxel = resampling_factor / (time_spacing_in_ms/1000 * target_size)
-    small_index = int(np.minimum((cutoff_highpass / single_voxel), target_size/2.0))
-    large_index = int(np.minimum((cutoff_lowpass / single_voxel), target_size/2.0))
+    large_index = np.floor(cutoff_lowpass_in_Hz/delta_f) # floor in order to ignore all frequencies above given lowpass cutoff
+    small_index = np.ceil(cutoff_highpass_in_Hz/delta_f) # ceil in order to ignore all frequencies below given highpass cutoff
+    large_index = int(np.clip(large_index, 0, len(frequencies)-1)) # limit by Nyquist frequency index
+    small_index = int(np.clip(small_index, 0, len(frequencies)-1)) # limit by Nyquist frequency index
 
     # construct bandpass filter given the cutoff values with tukey window in negative and positive frequencies
```
[the remark this answers is not in the document]
Comment (cbender98) [reply simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:199]: updatet this in the latest commit

**Review by TomTomRixRix — APPROVED** [review]
looks good to me
